### PR TITLE
[serve] Async controller

### LINKED ISSUE #12284: [serve] [autoscaling] Refactor controller to use async event loop

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
* Controller's API  *immediately* returns future
* Controller completes futures after the time 

## Related issue number
Closes  #12284

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -188,11 +189,38 @@ def get_backend_configs(self) -> Dict[BackendTag, BackendConfig]:
...
+        existing_goal = self.goals.get(backend_tag)
         self.backends[backend_tag] = backend_info
+        if not backend_info:
+            del self.backends[backend_tag]
+        self.goals[backend_tag] = goal_id
+        return existing_goal
+
+    def completed_goals(
+            self,
+            backend_replicas: Dict[BackendTag, Dict[ReplicaTag, ActorHandle]]
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:204]: current_state

File: python/ray/serve/controller.py [hunk 2]
```diff
@@ -757,12 +800,30 @@ async def do_autoscale(self) -> None:
...
+
     async def run_control_loop(self) -> None:
+        start_time = time.time()
         while True:
             await self.do_autoscale()
             async with self.write_lock:
                 self.http_state.update()
-
+                delta_workers = await self.\
+                    actor_reconciler.backend_control_loop(start_time)
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:816]: update_backend? or something that signifies that only one iteration is run.

File: python/ray/serve/controller.py [hunk 3]
```diff
@@ -923,8 +986,17 @@ async def delete_endpoint(self, endpoint: str) -> UUID:
...
             return return_uuid
 
+    async def modify_backend_goal(self, backend_tag: BackendTag,
+                                  backend_info: BackendInfo,
+                                  new_id: GoalId) -> None:
+        # NOTE(ilr) Must checkpoint after doing this!
+        existing_id_to_set = self.backend_state.add_or_modify_backend_goal(
+            backend_tag, backend_info, new_id)
+        self.set_goal_id(existing_id_to_set)
+
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:999]: Consolidate into modify or create new.

File: python/ray/serve/controller.py [hunk 4]
```diff
@@ -965,19 +1038,23 @@ async def create_backend(self, backend_tag: BackendTag,
             return_uuid = self._create_event_with_result({
                 backend_tag: backend_info
             })
+
+            await self.modify_backend_goal(backend_tag, backend_info,
+                                           return_uuid)
+
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:1044]: Delete
```suggestion

```

File: python/ray/serve/controller.py (hunk 3 above)
Comment (ijrsvt) [reply python/ray/serve/controller.py:999]: Declarative name: `set backend goal` (handles existing)

File: python/ray/serve/controller.py [hunk 5]
```diff
@@ -450,27 +477,39 @@ async def _check_currently_stopping_replicas(self) -> int:
...
+        num_stopping = len(self.currently_stopping_replicas)
         need_to_continue = True
         num_pending_starts, num_pending_stops = 0, 0
         while need_to_continue:
-            if time.time() - prev_warning > REPLICA_STARTUP_TIME_WARNING_S:
-                prev_warning = time.time()
-                delta = time.time() - start
+            if int(time.time() -
+                   start_time) % REPLICA_STARTUP_TIME_WARNING_S == 0:
+                delta = time.time() - start_time
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:512]: Switch to a per-goal start time.

File: python/ray/serve/controller.py [hunk 6]
```diff
@@ -450,27 +477,39 @@ async def _check_currently_stopping_replicas(self) -> int:
...
-    async def backend_control_loop(self):
-        start = time.time()
-        prev_warning = start
+    async def backend_control_loop(self, start_time: float) -> bool:
+        """Returns whether the number of backends has changed."""
+        num_starting = len(self.currently_starting_replicas)
+        num_stopping = len(self.currently_stopping_replicas)
         need_to_continue = True
         num_pending_starts, num_pending_stops = 0, 0
         while need_to_continue:
```
Comment (ijrsvt) [inline python/ray/serve/controller.py:509]: Remove looping here.

**Review by edoakes — APPROVED** [review]
Looks good. Can clean it up a bit by finishing the refactor w.r.t. actor reconciler, but as we discussed offline let's merge it as-is first once tests are passing.

**Pull request merged**